File: src/python/tree/103.binary-tree-zigzag-level-order-traversal.py
from typing import List, Optional
from lc_test_utils import TreeNode

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        res = []
        q = [root]
        left_to_right = True
        while q:
            level = []
            for _ in range(len(q)):
                node = q.pop(0)
                level.append(node.val)
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
            
            if not left_to_right:
                level.reverse()
            res.append(level)
            left_to_right = not left_to_right
        return res

File: src/go/tree/103.binary-tree-zigzag-level-order-traversal.go
package main

/*
 * @lc app=leetcode id=103 lang=golang
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 *
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/description/
 *
 * Given the root of a binary tree, return the zigzag level order traversal of
 * its nodes' values. (i.e., from left to right, then right to left for the
 * next level and alternate between).
 *
 * Example 1:
 * Input: root = [3,9,20,null,null,15,7]
 * Output: [[3],[20,9],[15,7]]
 *
 * Example 2:
 * Input: root = [1]
 * Output: [[1]]
 *
 * Example 3:
 * Input: root = []
 * Output: []
 *
 * Constraints:
 * - The number of nodes in the tree is in the range [0, 2000].
 * - -100 <= Node.val <= 100
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
	res := [][]int{}
	if root == nil {
		return res
	}
	q := []*TreeNode{root}
	leftToRight := true
	for len(q) > 0 {
		sz := len(q)
		lvl := make([]int, sz)
		for i := 0; i < sz; i++ {
			n := q[0]
			q = q[1:]
			idx := i
			if !leftToRight {
				idx = sz - 1 - i
			}
			lvl[idx] = n.Val
			if n.Left != nil {
				q = append(q, n.Left)
			}
			if n.Right != nil {
				q = append(q, n.Right)
			}
		}
		res = append(res, lvl)
		leftToRight = !leftToRight
	}
	return res
}
